src/monodroid/jni/monodroid-glue.cc

```diff
@@ -1498,6 +1570,30 @@ MonodroidRuntime::Java_mono_android_Runtime_initInternal (JNIEnv *env, jclass kl
 		utils.set_world_accessable (counters_path);
 	}
 
+	void *api_dso_handle = nullptr;
+#if defined (WINDOWS) || defined (APPLE_OS_X)
+	const char *my_location = get_my_location ();
+	if (my_location != nullptr) {
+		simple_pointer_guard<char, false> dso_path (utils.path_combine (my_location, API_DSO_NAME));
+		log_info (LOG_DEFAULT, "Attempting to load %s", dso_path.get ());
+		api_dso_handle = dlopen (dso_path.get (), RTLD_NOW);
+#if defined (APPLE_OS_X)
+		delete[] my_location;
+#else
+		free (static_cast<void*>(const_cast<char*>(my_location))); // JI allocates with `calloc`
+#endif
+	}
+
+	if (api_dso_handle == nullptr) {
+		log_info (LOG_DEFAULT, "Attempting to load %s with \"bare\" dlopen", API_DSO_NAME);
+		api_dso_handle = dlopen (API_DSO_NAME, RTLD_NOW);
+	}
+
+	if (api_dso_handle == nullptr)
+#endif
```

Comment:
This *looks* weird, because of the indentation. I would suggest moving this `#endif` *above* the `if`, which means the Android build source would be:

```cpp
void *api_dso_handle = nullptr;
if (api_dsl_handle == nullptr)
    api_dso_handle = androidSystem.load_dso_from_any_directories (API_DSO_NAME, RTLD_NOW);
```

Which *looks* better.  Arguably there's an "extra null check" in here, but the compiler should be able to optimize it away anyway…